ergo-lib/src/wallet/ext_secret_key.rs

```diff
@@ -0,0 +1,267 @@
+            ),
+            // m/0'/1/2'/2
+            Bip32Vector::new(
+                "2",
+                "0f479245fb19a38a1954c5c7c0ebab2f9bdfd96a17563ef28a6a4b1a2a764ef4",
+            ),
+            // m/0'/1/2'/2/1000000000
+            Bip32Vector::new(
+                "1000000000",
+                "471b76e389e528d6de6d816857e012c5455051cad6660850e58372a6c3e6e7c8",
+            ),
+        ];
+        let secret_key =
+            base16::decode(b"e8f32e723decf4051aefac8e2c93c9c5b214313817cdb01a1494b917c8436b35")
+                .unwrap();
+        let chain_code =
+            base16::decode(b"873dff81c02f525623fd1fe5167eac3a55a049de3d314bb42ee227ffed37d508")
+                .unwrap();
+        let mut ext_secret_key = ExtSecretKey::new(
+            secret_key.try_into().unwrap(),
+            chain_code.try_into().unwrap(),
+            DerivationPath::master_path(),
+        )
+        .unwrap();
+
+        for v in vectors {
+            ext_secret_key = ext_secret_key.derive(v.next_index).unwrap();
+            assert_eq!(ext_secret_key.secret_key_bytes(), v.expected_secret_key);
+        }
+    }
+
+    #[test]
+    fn bip32_test_vector2() {
+        let vectors = vec![
+            // m/0
+            Bip32Vector::new(
+                "0",
+                "abe74a98f6c7eabee0428f53798f0ab8aa1bd37873999041703c742f15ac7e1e",
+            ),
+            // m/0/2147483647'
+            Bip32Vector::new(
+                "2147483647'",
+                "877c779ad9687164e9c2f4f0f4ff0340814392330693ce95a58fe18fd52e6e93",
+            ),
+            // m/0/2147483647'/1
+            Bip32Vector::new(
+                "1",
+                "704addf544a06e5ee4bea37098463c23613da32020d604506da8c0518e1da4b7",
+            ),
+            // m/0/2147483647'/1/2147483646'
+            Bip32Vector::new(
+                "2147483646'",
+                "f1c7c871a54a804afe328b4c83a1c33b8e5ff48f5087273f04efa83b247d6a2d",
+            ),
+            // m/0/2147483647'/1/2147483646'/2
+            Bip32Vector::new(
+                "2",
+                "bb7d39bdb83ecf58f2fd82b6d918341cbef428661ef01ab97c28a4842125ac23",
+            ),
+        ];
+        let secret_key =
+            base16::decode(b"4b03d6fc340455b363f51020ad3ecca4f0850280cf436c70c727923f6db46c3e")
+                .unwrap();
+        let chain_code =
+            base16::decode(b"60499f801b896d83179a4374aeb7822aaeaceaa0db1f85ee3e904c4defbd9689")
+                .unwrap();
+        let mut ext_secret_key = ExtSecretKey::new(
+            secret_key.try_into().unwrap(),
+            chain_code.try_into().unwrap(),
+            DerivationPath::master_path(),
+        )
+        .unwrap();
+
+        for v in vectors {
+            ext_secret_key = ext_secret_key.derive(v.next_index).unwrap();
+            assert_eq!(ext_secret_key.secret_key_bytes(), v.expected_secret_key);
+        }
+    }
```

Comment:
Could you add test vectors from the node wallet implementation from https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKeySpec.scala#L18-L50
Just to be extra sure that they are equivalent.

Reply:
Good call on this, it made me realize our extended key interfaces might be a bit off compared to Scala.

See 8b8e9ff , the main thing being what was previously `derive` should be `child`, there should be a `derive` method that takes a full path instead of an index.

If you agree with the above I'd be happy to fix-up `ext_pub_key` and maybe extract things out into a `ext_key` trait like in node: https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/ExtendedKey.scala#L16

Reply:
> Good call on this, it made me realize our extended key interfaces might be a bit off compared to Scala.
> 
> See [here](https://github.com/ergoplatform/sigma-rust/commit/8b8e9ff85889d1771fbb024c18f8486b5cc1f618#diff-cc91791574e7a77dd357ac1364ee2c7bf71ace4a776d91b964cd7b774dc20344R102-R149), the main thing being what was previously `derive` should be `child`, there should be a `derive` method that takes a full path instead of an index.

Good catch on `child`! I don't see a `derive` with full path in node wallet but it's up to you.
 
> If you agree with the above I'd be happy to fix-up `ext_pub_key` and maybe extract things out into a `ext_key` trait like in node wallet: https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/ExtendedKey.scala#L16

I agree. I don't see `ExtendedKey` trait being used in the code, but feel free to implement it if you find it useful.

Reply:
> Good catch on child! I don't see a derive with full path in node wallet but it's up to you.

My Scala is bad but I was thinking that's what this was doing: https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/ExtendedKey.scala#L37

Which is covered by these tests in the original link (https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKeySpec.scala#L37-L50)

Reply:
> > Good catch on child! I don't see a derive with full path in node wallet but it's up to you.
> 
> My Scala is bad but I was thinking that's what this was doing, or at least the difference in the paths: https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/ExtendedKey.scala#L37
> 
> Which is covered by these tests in the original link (https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKeySpec.scala#L37-L50)

Oh yes, you're right, I missed it. I was looking if the trait is used as generic type somewhere to accommodate both.